d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -239,6 +295,71 @@ XdsClient.D2NodeMapResourceWatcher getUriResourceWatcher(String clusterName)
     return new UriPropertiesResourceWatcher(clusterName);
   }
 
+  XdsClient.D2SymlinkNodeResourceWatcher getSymlinkResourceWatcher(String symlinkName)
+  {
+    return new XdsClient.D2SymlinkNodeResourceWatcher()
+    {
+      @Override
+      public void onChanged(String resourceName, XdsClient.D2SymlinkNodeUpdate update)
+      {
+        // update maps between symlink name and actual node name, listen to the actual node
+        String actualResourceName = update.getNodeData().getMasterClusterNodePath();
+        if (resourceName.contains(D2_CLUSTER_NODE_PREFIX))
+        {
+          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_CLUSTER_NODE_PREFIX);
+          updateSymlinkAndActualNodeMaps(symlinkName, actualNodeName);
+          listenToCluster(actualNodeName);
+        }
+        else
+        {
+          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_URI_NODE_PREFIX);
+          updateSymlinkAndActualNodeMaps(symlinkName, actualNodeName);
+          listenToUris(actualNodeName);
+        }
+      }
+
+      @Override
+      public void onError(Status error)
+      {
+        notifyAvailabilityChanges(false);
+      }
+
+      @Override
+      public void onReconnect()
+      {
+        notifyAvailabilityChanges(true);
+      }
+    };
+  }
+
+  private void updateSymlinkAndActualNodeMaps(String symlinkName, String actualNodeName)
+  {
+    // update symlinkToActualNode map.
+    _symlinkToActualNode.compute(symlinkName, (k, v) -> {
+      // If existing actual node differs from new node, remove the existing entry in actualNodeToSymlink map.
+      if (v != null && !v.equals(actualNodeName))
+      {
+        _actualNodeToSymlink.remove(v);
+      }
+      return actualNodeName;
+    });
+
+    _actualNodeToSymlink.put(actualNodeName, symlinkName);
+  }
```

Comment:
I think this is too likely to become racy. A read that comes in between `_symlinkToActualNode.compute` and `_actualNodeToSymlink.put` can read an incorrect value.

IMO, this would be safer:
```java
private final Object _symlinkToActualNodeLock = new Object();
private final HashBiMap<String, String> _symlinkToActualNode = HashBiMap.create();

private void updateSymlinkAndActualNodeMaps(String symlinkName, String actualNodeName) {
  synchronized (_symlinkToActualNodeLock) {
    _symlinkToActualNode.put(symlinkName, actualNodeName);
  }
}

private String getSymlink(String actualNodeName) {
  synchronized (_symlinkToActualNodeLock) {
    return _symlinkToActualNode.inverse().get(actualNodeName);
  }
}
```
Alternatively, if `HashBiMap` from Guava isn't available in the rest.li code, you can still achieve the same thing with the two maps but you still must lock both

Reply:
yeah, I thought about that too. But the xDSClientImpl is designed to have only one thread processing the response, which means only one resource update will happen at a time (only one watcher's onChanged will be called a time). So a read of the actualNodeToSymlink (inside the cluster resource watcher) will never come in when an update of the maps is ongoing.

Reply:
`HashBiMap` seems like an elegant solution, but `_actualNodeToSymlink.put(actualNodeName, symlinkName);` could also be moved inside the compute method lambda?

Reply:
I ended up making it thread-safe with HashBiMap, just in case we need to add threads to xDSClient in the future.